patches/tModLoader/Terraria.ModLoader/Logging.cs

```diff
@@ -188,27 +188,38 @@ public static class Logging
 			if (handlerActive.Value)
 				return;
 
+			bool oom = args.Exception is OutOfMemoryException;
+
+			//In case of OOM, unload the Main.tile array and do immediate garbage collection.
+			//If we don't do this, there will be a big chance that this method will fail to even quit the game, due to another OOM exception being thrown.
+			if (oom) {
+				Main.tile = null;
+
+				GC.Collect();
+			}
+
 			try {
 				handlerActive.Value = true;
 
-				if (args.Exception == previousException ||
-					args.Exception is ThreadAbortException ||
-					ignoreSources.Contains(args.Exception.Source) ||
-					ignoreMessages.Any(str => args.Exception.Message?.Contains(str) ?? false) ||
-					ignoreThrowingMethods.Any(str => args.Exception.StackTrace?.Contains(str) ?? false))
-					return;
+				if (!oom)
+					if (args.Exception == previousException ||
+						args.Exception is ThreadAbortException ||
+						ignoreSources.Contains(args.Exception.Source) ||
+						ignoreMessages.Any(str => args.Exception.Message?.Contains(str) ?? false) ||
+						ignoreThrowingMethods.Any(str => args.Exception.StackTrace?.Contains(str) ?? false))
+						return;
 
 				var stackTrace = new StackTrace(true);
 				PrettifyStackTraceSources(stackTrace.GetFrames());
 				var traceString = stackTrace.ToString();
 
-				if (ignoreContents.Any(traceString.Contains))
+				if (!oom && ignoreContents.Any(traceString.Contains))
 					return;
 
 				traceString = traceString.Substring(traceString.IndexOf('\n'));
 				var exString = args.Exception.GetType() + ": " + args.Exception.Message + traceString;
 				lock (pastExceptions) {
-					if (!pastExceptions.Add(exString))
+					if (!pastExceptions.Add(exString) && !oom)
```

Comment:
redundant, can't have 2 ooms in a row

Reply:
We indeed can and did in the past. That exact return had to be the *second* reason why the game would continue to look semi-fine after encountering OOM and failing to show the error message. But sure, I'll remove the check, hoping that Main.tile unloading will always free enough memory.